src/logfile_values.py

```diff
@@ -7,42 +7,298 @@
 """
 logfile_values.py -- a python script to find metrics values in log file
 
-This script is using last line of log file to get metric value by column number
+This script can retrive last line values by column number:
 
 python logfile_values.py --metric "metric1:1" "metric2:2" ...
+
+And aggregate data by time period with different functions:
+
+median, mean, sum, min, max, count, frequency, speed (??)
+
+count_100 - counts values > 100
+count_100_percentage - estimates percentage of values > 100
+
+python logfile_values.py  --metric "metric1:1:mean:1d" --metric "metric2:3:count:60s"
+
 """
+import re
+import time
+import os
+import gzip
+import logging
 
 from argparse import ArgumentParser, ArgumentTypeError
-from time import time
+from datetime import timedelta
 
 
 class Metric:
     def __init__(self, arg):
         if ':' not in arg:
             raise ArgumentTypeError('Argument must have ":"')
-        self.name, column = arg.split(':', 1)
-        if not column.isdecimal():
-            raise ArgumentTypeError('Column must be a number')
-        self.column = int(column)
+        parts = arg.split(':')
+        if len(parts) != 4 and len(parts) != 2:
+            raise ValueError('Wrong number of options')
+
+        # Accepts 2 or 4 options. If 2 then function and period are NULL
+        parts += [None] * (4 - len(parts))
+        self.name, column, function, period = parts
+        if period:
+            pattern = re.compile("^\d+[A-Za-z]{1,3}$")
+            if not pattern.match(period):
+                raise ArgumentTypeError('Period must have number and unit')
+
+        self.column = column
+        self.function = function
+        self.period = period
+        # Here is container for metric values
+        self.values = []
+        self.last_value = 0
+        self.start_timestamp = 0
+        self.end_timestamp = 0
+        self.now = int(time.time())
+
+    def get_timeshift(self):
+        if self.period:
+            value = int(" ".join(re.findall("\d+", self.period.lower())))
+            unit = " ".join(re.findall("[a-z]+", self.period.lower()))
+            if unit == 's':
+                return timedelta(seconds=value).total_seconds()
+            if unit == 'min':
+                return timedelta(minutes=value).total_seconds()
+            if unit == 'h':
+                return timedelta(hours=value).total_seconds()
+            if unit == 'd':
+                return timedelta(days=value).total_seconds()
```

Comment:
Regexp is overkill to parse something so simple.  How about something like this:
```
units = [
    ('s', 1)
    ('min', 60),
    ('h', 60 * 60),
]

for unit, mul in units:
    if self.period.lower().endswith(unit):
        return int(self.period[:len(unit)].strip()) * mul
```

Reply:
+1